File: shared/scanPack.ts
import type { Board } from 'chessops/board';
import { Chess } from 'chessops/chess';
import { parseSan } from 'chessops/san';
import { squareFile } from 'chessops/util';
import type { Role } from 'chessops/types';
import { hashSetup } from './zobrist.ts';

/**
 * The packed scan-index: one blob per game, everything a deep scan
 * tests, precomputed — so a future scanner reads bytes instead of
 * parsing SAN and replaying (the whole cost of today's scan, measured
 * 12.7 s JS / 1.3 s native per Elite month).
 *
 * This module is the SPEC as much as the implementation. The Rust
 * indexer (native/src/scan_pack.rs) must emit byte-identical blobs —
 * pinned by fixtures exported from this file into the goldens — and a
 * blob, once written, is read by whichever side scans it, so the layout
 * below can only ever be extended behind a version bump, never changed.
 *
 * ## Layout (version 1, all integers little-endian)
 *
 * ```
 * u16          npos   — positions encoded: replayed plies + 1
 * npos × u32   key32  — per position, the low 32 bits of the UNSIGNED
 *                       zobrist hash (shared/zobrist.ts, before the
 *                       signed db-key conversion), position 0 first
 * npos × u8    pawns8 — per position, the pawn-files hash below
 * (npos−1) × u8 event — per ply, what the move did to the material
 * ```
 *
 * The replay follows the plies index's own stopping rules — the first
 * SAN token that fails to parse or is illegal ends the stream, keeping
 * everything before it — but has NO ply cap: the pack exists precisely
 * for the depths the plies table stops short of. A 32-bit key is a
 * prefilter, not an answer: a scanner that matches one must verify the
 * hit by replaying that one game, which is cheap because hits are few.
 *
 * ## The pawn-files hash
 *
 * One byte discriminating pawn STRUCTURE, for the ladder's pawns and
 * files rungs: without it their candidates were every game whose piece
 * counts matched, and verification dominated (measured 112 s over five
 * million games; the whole point of the pack is not replaying). The
 * hash digests the files-rung projection — per file, per side, how
 * many pawns — so it is exact for the files rung up to its 8 bits, and
 * a sound prefilter for the pawns rung (same placement implies same
 * file counts). Defined arithmetic, not a library: h starts at 5, and
 * for each of the 16 counts v (white files a…h, then black a…h),
 * h = (h * 33 + v) mod 256.
 *
 * ## The event byte
 *
 * `captured | promoted << 3`, with roles numbered none=0, pawn=1,
 * knight=2, bishop=3, rook=4, queen=5 (kings are never captured, and a
 * promotion is never to a pawn or king). The captured side is the side
 * that did NOT move — ply parity says which. From these events alone a
 * scanner reconstructs every position's per-piece counts, which is what
 * the material hunt and the ladder's cheap gates test.
 *
 * Deriving the two fields is specified so both implementations agree
 * without sharing a chess library:
 * - promoted: the move's own promotion role, none otherwise.
 * - captured: the role of the ENEMY piece on the destination square —
 *   an own piece there is castling (king takes rook encoding), which
 *   captures nothing; a pawn landing on an EMPTY square while changing
 *   file is en passant and captures a pawn.
 */
export const SCAN_PACK_VERSION = 1;

/** The meta key naming the version of the packs a database carries —
    absent means no packs (or packs with holes: an append refuses to
    leave a partially packed file, see the indexers). */
export const SCAN_PACK_META = 'scan_pack';

const ROLE_CODE: Record<Role, number> = {
  pawn: 1,
  knight: 2,
  bishop: 3,
  rook: 4,
  queen: 5,
  king: 0, // never captured, never promoted to — encoded as none
};

const MASK32 = 0xffffffffn;

/** The pawn-files hash of a board — see the header for the exact
    arithmetic, which the Rust twin repeats digit for digit. */
export function pawnFilesHash(board: Board): number {
  const files = new Array<number>(16).fill(0);
  for (const square of board.pawn.intersect(board.white)) {
    files[squareFile(square)] = (files[squareFile(square)] ?? 0) + 1;
  }
  for (const square of board.pawn.intersect(board.black)) {
    files[8 + squareFile(square)] = (files[8 + squareFile(square)] ?? 0) + 1;
  }
  let h = 5;
  for (const value of files) h = (h * 33 + value) & 0xff;
  return h;
}

/** Encode one game's movetext into its pack. Replays with the same
    chessops pipeline as every other consumer, full depth. */
export function encodeScanPack(moves: string): Uint8Array {
  const pos = Chess.default();
  const keys: number[] = [Number(hashSetup(pos.toSetup()) & MASK32)];
  const pawns: number[] = [pawnFilesHash(pos.board)];
  const events: number[] = [];
  for (const san of moves.split(' ')) {
    const move = parseSan(pos, san);
    if (!move) break;
    let event = 0;
    if ('from' in move) {
      const mover = pos.board.get(move.from);
      const target = pos.board.get(move.to);
      if (target && target.color !== pos.turn) {
        event |= ROLE_CODE[target.role];
      } else if (
        !target &&
        mover?.role === 'pawn' &&
        squareFile(move.from) !== squareFile(move.to)
      ) {
        event |= ROLE_CODE.pawn; // en passant
      }
      if (move.promotion) event |= ROLE_CODE[move.promotion] << 3;
    }
    pos.play(move);
    events.push(event);
    keys.push(Number(hashSetup(pos.toSetup()) & MASK32));
    pawns.push(pawnFilesHash(pos.board));
  }
  const npos = keys.length;
  const pack = new Uint8Array(2 + 5 * npos + events.length);
  const view = new DataView(pack.buffer);
  view.setUint16(0, npos, true);
  keys.forEach((key, at) => view.setUint32(2 + 4 * at, key, true));
  pawns.forEach((hash, at) => {
    pack[2 + 4 * npos + at] = hash;
  });
  events.forEach((event, at) => {
    pack[2 + 5 * npos + at] = event;
  });
  return pack;
}

/** A decoded pack, for tests and the scanner to come. */
export interface ScanPack {
  keys: number[];
  pawns: number[];
  events: number[];
}

/** Decode a pack, or null when the bytes are not a well-formed v1 blob
    — truncated, oversized, or an impossible header. */
export function decodeScanPack(pack: Uint8Array): ScanPack | null {
  if (pack.length < 2) return null;
  const view = new DataView(pack.buffer, pack.byteOffset, pack.byteLength);
  const npos = view.getUint16(0, true);
  if (npos < 1 || pack.length !== 2 + 5 * npos + (npos - 1)) return null;
  const keys: number[] = [];
  for (let at = 0; at < npos; at += 1) keys.push(view.getUint32(2 + 4 * at, true));
  const pawns: number[] = [];
  for (let at = 0; at < npos; at += 1) pawns.push(pack[2 + 4 * npos + at]!);
  const events: number[] = [];
  for (let at = 0; at < npos - 1; at += 1) events.push(pack[2 + 5 * npos + at]!);
  return { keys, pawns, events };
}

File: native/src/scan_pack.rs
//! Port of `shared/scanPack.ts`: the packed scan-index blob, one per
//! game. That file is the spec — layout, event byte, stopping rules —
//! and the goldens pin this port to its bytes: `encode_scan_pack` must
//! produce exactly what the JS encoder wrote for the same movetext,
//! because whichever side scans a database reads blobs the other may
//! have written.

use shakmaty::{san::SanPlus, Board, Chess, Move, Position, Role};

use crate::zobrist::hash_position;

pub const SCAN_PACK_VERSION: u32 = 1;
pub const SCAN_PACK_META: &str = "scan_pack";

/// The pawn-files hash — the spec's own arithmetic (h = 5, then
/// h = (h * 33 + v) mod 256 over white files a…h then black a…h).
fn pawn_files_hash(board: &Board) -> u8 {
    let mut files = [0u32; 16];
    for sq in board.pawns() & board.white() {
        files[u32::from(sq.file()) as usize] += 1;
    }
    for sq in board.pawns() & board.black() {
        files[8 + u32::from(sq.file()) as usize] += 1;
    }
    let mut h: u32 = 5;
    for value in files {
        h = (h.wrapping_mul(33).wrapping_add(value)) & 0xff;
    }
    h as u8
}

/// Role numbering of the event byte: none=0, pawn=1 … queen=5. Kings
/// are never captured and never promoted to.
fn role_code(role: Role) -> u8 {
    match role {
        Role::Pawn => 1,
        Role::Knight => 2,
        Role::Bishop => 3,
        Role::Rook => 4,
        Role::Queen => 5,
        Role::King => 0,
    }
}

/// Encode one game's movetext, full depth, the plies replay's stopping
/// rules (first unparseable or illegal SAN ends the stream).
/** Per-piece counts in p,n,b,r,q order per side, plus totals with the
    kings — the envelope's raw material. */
fn board_counts(board: &Board) -> ([u8; 5], [u8; 5], u8, u8) {
    let sets = [
        board.pawns(),
        board.knights(),
        board.bishops(),
        board.rooks(),
        board.queens(),
    ];
    let mut w = [0u8; 5];
    let mut b = [0u8; 5];
    let mut w_tot = 1u8;
    let mut b_tot = 1u8;
    for (at, set) in sets.iter().enumerate() {
        w[at] = (*set & board.white()).count() as u8;
        b[at] = (*set & board.black()).count() as u8;
        w_tot += w[at];
        b_tot += b[at];
    }
    (w, b, w_tot, b_tot)
}

pub fn encode_scan_pack(moves: &str) -> Vec<u8> {
    let mut pos = Chess::default();
    let mut keys: Vec<u32> = vec![hash_position(&pos) as u32];
    let mut pawns: Vec<u8> = vec![pawn_files_hash(pos.board())];
    let mut events: Vec<u8> = Vec::new();
    let mut min = [8u8, 2, 2, 2, 1, 8, 2, 2, 2, 1];
    let mut max = [8u8, 2, 2, 2, 1, 8, 2, 2, 2, 1];
    let mut min_w_tot = 16u8;
    let mut min_b_tot = 16u8;
    for token in moves.split(' ') {
        let Ok(san) = token.parse::<SanPlus>() else {
            break;
        };
        let Ok(m) = san.san.to_move(&pos) else {
            break;
        };
        // shakmaty's move variants carry exactly the spec's derivation:
        // Normal{capture} is the enemy piece on the destination,
        // EnPassant captures a pawn, Castle captures nothing.
        let mut event = m.capture().map_or(0, role_code);
        if let Move::Normal {
            promotion: Some(role),
            ..
        } = m
        {
            event |= role_code(role) << 3;
        }
        pos.play_unchecked(m);
        events.push(event);
        keys.push(hash_position(&pos) as u32);
        pawns.push(pawn_files_hash(pos.board()));
        let (w, b, w_tot, b_tot) = board_counts(pos.board());
        for at in 0..5 {
            min[at] = min[at].min(w[at]);
            max[at] = max[at].max(w[at]);
            min[5 + at] = min[5 + at].min(b[at]);
            max[5 + at] = max[5 + at].max(b[at]);
        }
        min_w_tot = min_w_tot.min(w_tot);
        min_b_tot = min_b_tot.min(b_tot);
    }
    let npos = keys.len();
    let mut pack = Vec::with_capacity(13 + 6 * npos);
    pack.extend_from_slice(&(npos as u16).to_le_bytes());
    pack.push(min_w_tot);
    pack.push(min_b_tot);
    for at in 0..10 {
        pack.push((min[at] << 4) | max[at]);
    }
    for key in keys {
        pack.extend_from_slice(&key.to_le_bytes());
    }
    pack.extend_from_slice(&pawns);
    pack.extend_from_slice(&events);
    pack
}
